src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,554 @@
+#include "postgres.h"
+
+#include "funcapi.h"
+#include "libpq-fe.h"
+#include "miscadmin.h"
+
+#include "access/hash.h"
+#include "distributed/backend_data.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/metadata_cache.h"
+#include "storage/proc.h"
+#include "utils/builtins.h"
+#include "utils/hsearch.h"
+#include "utils/timestamp.h"
+
+
+/*
+ * Describes an edge in a waiting-for graph of locks.  This isn't used for
+ * deadlock-checking directly, but to gather the information necessary to
+ * do so.
+ *
+ * The datatypes here are a bit looser than strictly necessary, because
+ * they're transported as the return type from an SQL function.
+ */
+typedef struct WaitEdge
+{
+	int waitingPid;
+	int waitingNodeId;
+	int64 waitingTransactionNum;
+	TimestampTz waitingTransactionStamp;
+
+	int blockingPid;
+	int blockingNodeId;
+	int64 blockingTransactionNum;
+	TimestampTz blockingTransactionStamp;
+
+	/* blocking transaction is also waiting on a lock */
+	bool isBlockingXactWaiting;
+} WaitEdge;
+
+
+/*
+ * WaitGraph represent a graph of wait edges as an adjacency list.
+ */
+typedef struct WaitGraph
+{
+	int localNodeId;
+	size_t numAllocated;
+	size_t numUsed;
+	WaitEdge *edges;
+} WaitGraph;
+
+
+/*
+ * PROCStack is a stack of PGPROC pointers used to perform a depth-first search
+ * through the lock graph.
+ */
+typedef struct PROCStack
+{
+	int procCount;
+	PGPROC **procs;
+} PROCStack;
+
+
+static void ReturnWaitGraph(WaitGraph *waitGraph, FunctionCallInfo fcinfo);
+static WaitGraph * BuildWaitGraphForSourceNode(int sourceNodeId);
+static void LockLockData(void);
+static void UnlockLockData(void);
+static void AddEdgesForLockWaits(WaitGraph *waitGraph, PGPROC *waitingProc,
+								 PROCStack *remaining);
+static void AddEdgesForWaitQueue(WaitGraph *waitGraph, PGPROC *waitingProc,
+								 PROCStack *remaining);
+static void AddWaitEdge(WaitGraph *waitGraph, BackendData *waitingXactData,
+						PGPROC *waitingProc, BackendData *blockingXactData,
+						PGPROC *blockingProc, bool isBlockingProcessWaiting);
+static WaitEdge * AllocWaitEdge(WaitGraph *waitGraph);
+static bool IsProcessWaitingForLock(PGPROC *proc);
+static bool IsInDistributedTransaction(BackendData *backendData);
+static bool IsSameLockGroup(PGPROC *leftProc, PGPROC *rightProc);
+
+
+PG_FUNCTION_INFO_V1(dump_local_wait_edges);
+
+
+/*
+ * dump_local_wait_edges returns wait edges for distributed transactions
+ * running on the node on which it is called, which originate from the source node.
+ */
+Datum
+dump_local_wait_edges(PG_FUNCTION_ARGS)
+{
+	int32 sourceNodeId = PG_GETARG_INT32(0);
+
+	WaitGraph *waitGraph = BuildWaitGraphForSourceNode(sourceNodeId);
+	ReturnWaitGraph(waitGraph, fcinfo);
+
+	return (Datum) 0;
```

Comment:
The `return (Datum) 0;` seems to be the common pattern for SRFs in postgres, even though `PG_RETURN_VOID()` is the same. I presume it's because `PG_RETURN_VOID()` is meant for void functions.